### PR TITLE
refactor build fn to use a class for managing state

### PR DESCRIPTION
**Pull request by jakemac53**
closes https://github.com/dart-lang/build/issues/43

### CODE REVIEW

**Comment by jakemac53** [comment]
Essentially this just moves all top level functions to instance methods, and zone-local values to properties. I also simplified some argument signatures by adding additional properties.

File: lib/src/generate/build_impl.dart [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+import 'phase.dart';
+
+/// Class which manages running builds.
+class BuildImpl {
+  final AssetGraph _assetGraph;
+  final AssetReader _reader;
+  final AssetWriter _writer;
+  final PackageGraph _packageGraph;
+  final List<List<Phase>> _phaseGroups;
+  Map<String, Set<AssetId>> _inputsByPackage;
```
Comment (kegluneq) [inline lib/src/generate/build_impl.dart]: nit: Make this `final` & assign right away.

Change the assignment in `_initializeInputsByPackage` to a call to `clear()`

File: lib/src/generate/build_impl.dart [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+  bool _buildRunning = false;
+
+  BuildImpl(this._assetGraph, this._reader, this._writer, this._packageGraph,
+      this._phaseGroups);
+
+  /// Runs a build.
```
Comment (kegluneq) [inline lib/src/generate/build_impl.dart]: Consider adding an explanation of what is returned, something like:

```dart
/// Runs a build
/// 
/// The returned [Future] is guaranteed to complete with a [BuildResult]. If 
/// an exception is thrown by any phase, [BuildResult#status] will be set to 
/// [BuildStatus.Failure]. The exception and stack trace that caused the failure
/// will be available as [BuildResult#exception] and [BuildResult#stackTrace]
/// respectively.
```

File: lib/src/generate/build_impl.dart [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+      final groupOutputIds = <AssetId>[];
+      for (var phase in group) {
+        var inputs = _matchingInputs(phase.inputSets);
+        for (var input in inputs) {
+          for (var builder in phase.builders) {
+            var outputs = builder.declareOutputs(input);
+
+            groupOutputIds.addAll(outputs);
+            for (var output in outputs) {
+              if (tempInputsByPackage[output.package]?.contains(output) == true) {
```
Comment (kegluneq) [inline lib/src/generate/build_impl.dart]: Aw, I see why this needs to happen, but it makes me sad.

**Comment by kegluneq** [comment]
LGTM

File: lib/src/generate/build_impl.dart (hunk 2 above)
Comment (jakemac53) [inline lib/src/generate/build_impl.dart]: done, I copied your description cause I liked it :)

File: lib/src/generate/build_impl.dart (hunk 3 above)
Comment (jakemac53) [inline lib/src/generate/build_impl.dart]: Ya :(, at least its not hitting the file system twice now lol :)

**Pull request merged**